src/core/include/mp-units/bits/scaling.h

```diff
@@ -0,0 +1,229 @@
+//
+// Copyright (c) 2018 Mateusz Pusz
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+// IWYU pragma: private, include <mp-units/framework.h>
+#include <mp-units/bits/fixed_point.h>
+#include <mp-units/framework/customization_points.h>
+
+#ifndef MP_UNITS_IN_MODULE_INTERFACE
+#ifdef MP_UNITS_IMPORT_STD
+import std;
+#else
+#include <concepts>
+#endif
+#endif
+
+
+namespace mp_units {
+
+namespace detail {
+
+template<std::floating_point A, std::floating_point B>
+using minimal_floating_point_type =
+  std::conditional_t<(std::numeric_limits<A>::digits <= std::numeric_limits<B>::digits), A, B>;
+
+template<typename To, typename T>
+constexpr auto cast_integral(const T& value)
+{
+  if constexpr (std::is_integral_v<T>) {
+    return static_cast<To>(value);
+  } else {
+    return value;
+  }
+}
+
+template<typename T>
+struct floating_point_scaling_factor_type {
+  // fallback implementation for types with a `value_type` nested type
+  using type =
+    std::enable_if_t<!std::is_same_v<value_type_t<T>, T>, floating_point_scaling_factor_type<value_type_t<T>>>::type;
+};
+
+template<std::floating_point T>
+struct floating_point_scaling_factor_type<T> : std::type_identity<T> {};
+
+template<std::integral T>
+struct floating_point_scaling_factor_type<T> : std::common_type<T, float> {};
+
+template<typename T>
+  requires requires { typename scaling_traits<T>::floating_point_scaling_factor_type; }
+struct floating_point_scaling_factor_type<T> :
+    std::type_identity<typename scaling_traits<T>::floating_point_scaling_factor_type> {};
+
+
+template<Magnitude auto M>
+struct floating_point_scaling_impl {
+  static constexpr Magnitude auto num = _numerator(M);
+  static constexpr Magnitude auto den = _denominator(M);
+  template<typename T>
+  static constexpr T ratio = [] {
+    using U = long double;
+    return static_cast<T>(_get_value<U>(M));
+  }();
+
+  template<typename To, typename From>
```

Comment:
No. `floating_point_scaling_impl` is not intended as "an implementation for floating point types" - it is "an implementation as multiplication with a floating-point scaling factor". As mentioned above, this includes cases where just one of the two is a floating-point type.

Reply:
In such a case, `minimal_floating_point_type` in the first line of its implementation should also not be constrained; otherwise, the code will not compile with other types.